### Problem Set 1

#### Question 6 Python

You can turn a word into pig-Latin using the following two rules(simplified):
- If a word starts with a consonant, move that letter to the end and append 'ap'. For example, 'happy' becomes 'appyhay' and 'pencil' becomes 'encilpay'.
- If the word starts with a vowel, simply append 'way' to the end of the word. For example, 'enter' becomes 'enterway' and 'other' becomes 'otherway'. For our purposes, there are 5 vowels: a,e,i,o,u (so we count y as a consonant). Write a function pig() that takes a word (i.e. a string) as input and returns its pig-Latin form. Your function should still work if the input word contains upper case characters. Your output should always be the lowercase however.

#### Solution 6 Python

In [28]:
def pig(word):
    y = word.lower()
    words = y.split()
    vowels = ("a","e","i","o","u")
    if y[0] in vowels:
        for word in words:
            print(word + "way", end = " ")
    else:
        for word in words:
            print(word[1:] + word[0] + "ay", end = " ")
    print()
    
pig('happy')
pig('enter')


appyhay 
enterway 


#### Solution 6 Output

appyhay

enterway 

===============================================================================================================================

#### Question 7 Python

File bloodtype1.txt records blood-types of patients (A, B, AB, O OR OO) at a clinic. Write function bldcount() that reads the file with name 'name' and reports (i.e. prints) how many patients there are in each bloodtype.

#### Solution 7 Python

In [25]:
from varname import nameof

def bldcount(txt):
    with open(txt) as f:
        lines = f.read()
    bg = lines.split()
    
    global a,b,ab,o,oo
    a = 0
    b = 0
    ab = 0
    o = 0
    oo = 0
    for blood in bg:
        if blood == 'A' :
            a = a+1            
        elif blood == 'B':
            b = b+1
        elif blood == 'AB':
            ab = ab+1
        elif blood == 'O':
            o = o+1
        elif blood == 'OO':
            oo = oo+1
    print("There " + ('is ' if (a==1) else 'are ') + ('no' if (a==0) else str(a)) + (' patient' if (b==1) else ' patients') +" patients of blood type A")
    print("There " + ('is ' if (b==1) else 'are ' )+ ('no' if (b==0) else str(b)) + (' patient' if (b==1) else ' patients') +" of blood type B")
    print("There " + ('is ' if (ab==1) else 'are ' )+ ('no' if (ab==0) else str(ab)) + (' patient' if (b==1) else ' patients') +" patients of blood type AB")
    print("There " + ('is ' if (o==1) else 'are ' )+ ('no' if (o==0) else str(o)) + (' patient' if (b==1) else ' patients') +" patients of blood type O")
    print("There " + ('is ' if (oo==1) else 'are ' )+ ('no' if (oo==0) else str(oo)) + (' patient' if (b==1) else ' patients') +" patients of blood type OO")
    
bldcount('bloodtype1.txt')

There are 15 patients of blood type A
There is 1 patient of blood type B
There are 13 patients of blood type AB
There are 15 patients of blood type O
There are no patients of blood type OO


#### Solution 7 Output

There are 15 patients of blood type A

There is 1 patient of blood type B

There are 13 patients of blood type AB

There are 15 patients of blood type O

There are no patients of blood type OO

===============================================================================================================================

#### Question 8 Python

Write a function curconv() that takes as input:

1. a currency represented using a string (eg, 'JPY' for the Japanese Yen or 'EURO' for the Euro)
2. an amount

and then converts and returns the amount in US dollars.

The currency rates you need are stored in file currencies.txt

#### Solution 8 Python

In [23]:
currency = ""
res = int
def curconv(cur,amt):
    if cur == "AUD":
        res = 1.0345157*amt
        currency = "Australian Dollar"
    elif cur == "CHF":
        res =  1.0237414*amt
        currency = "Swiss Franc"
    elif cur == "CNY":
        res =  0.1550176*amt
        currency = "Chinese Yuan"
    elif cur == "DKK":
        res =  0.1651442*amt
        currency = "Danish Krone"
    elif cur == "EUR":
        res =  1.2296544*amt
        currency = "Euro"
    elif cur == "GBP":
        res =  1.5550989*amt
        currency = "British Pound"
    elif cur == "HKD":
        res =  0.1270207*amt
        currency = "Hong Kong Dollar"
    elif cur == "INR":
        res =  0.0177643*amt
        currency = "Indian Rupee"
    elif cur == "JPY":
        res =  0.01241401*amt
        currency = "Japanese Yen"
    elif cur == "MXN":
        res =  0.0751848*amt
        currency = "Mexican Peso"
    elif cur == "MYR":
        res =  0.3145411*amt
        currency = "Malaysian Ringgit"
    elif cur == "NOK":
        res =  0.1677063*amt
        currency = "Norwegian Krone"
    elif cur == "NZD":
        res =  0.8003591*amt
        currency = "New Zealand Dollar"
    elif cur == "PHP":
        res =  0.0233234*amt
        currency = "Philippine Peso"
    elif cur == "SEK":
        res =  0.148269*amt
        currency = "Swedish Krona"
    elif cur == "SGD":
        res =  0.788871*amt
        currency = "Singapore Dollar"
    elif cur == "THB":
        res =  0.0313789*amt
        currency = "Thai Baht"
        
    print (str(res))

curconv("EUR" , 100)
curconv("JPY" , 100)


122.96544
1.241401


#### Solution 8 Output

122.96544

1.241401


========================================================================================================================

#### Question 9 Python

Each of the following will cause an exception (an error). Identify what type of exception each will cause

a - Trying to add incompatible variables, as in adding 6 + 'a'

#### Solution 9 - a

In [3]:
print(6 + 'a');

TypeError: unsupported operand type(s) for +: 'int' and 'str'

#### Solution 9 - a Output

TypeError: unsupported operand type(s) for +: 'int' and 'str'


=========================================================================================================================

b- Referring to the 12th item of a list that has only 10 items

#### Solution 9 - b

In [4]:
list = [9, 2, 7, 1, 8, 4, 1, 3, 6, 4]
print(list[12]);

IndexError: list index out of range

#### Solution 9 - b Output

IndexError: list index out of range

=======================================================================================================================

c- Using a value that is out of range for a function's input, such as calling math.sqrt(-1.0)

#### Solution 9 - c

In [5]:
import math

math.sqrt(-1.0)

ValueError: math domain error

#### Solution 9 - c Output

ValueError: math domain error

========================================================================================================================

d - Using an undeclared variable, such as print(x) when x has not been defined

#### Solution 9 - d

In [27]:
print(X);

NameError: name 'X' is not defined

#### Solution 9 - d Output

NameError: name 'x' is not defined

========================================================================================================================

e - Trying to open a file that does not exist, such as mistyping the file name or looking in the wrong directory.

In [7]:
f = open("test.txt")
f.close()

FileNotFoundError: [Errno 2] No such file or directory: 'test.txt'

#### Solution 9 - e Output

FileNotFoundError: [Errno 2] No such file or directory: 'test.txt'

===============================================================================================================================

#### Question 10 Python

Assume that string of letters is already defined as 'abcdefghijklmnopqrstuvwxyz'. Write a function called frequencies() that takes a string as its only parameter, and returns a list of integers, showing the number of times each character appears in the text. Your function may ignore any characters that are not in letters

#### Solution 10 Python

In [8]:
def frequencies(word):
  alphabets = "abcdefghijklmnopqrstuvwqyz"
  counts = []
  for letter in alphabets:
    count = word.count(letter)
    counts.append(count)
  print(counts)   
    
frequencies("The quick red fox got bored and went home.")
frequencies("apple")

[1, 1, 1, 3, 5, 1, 1, 2, 1, 0, 1, 0, 1, 2, 4, 0, 1, 2, 0, 2, 1, 0, 1, 1, 0, 0]
[1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


#### Solution 10 Output

[1, 1, 1, 3, 5, 1, 1, 2, 1, 0, 1, 0, 1, 2, 4, 0, 1, 2, 0, 2, 1, 0, 1, 1, 0, 0]

[1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]